### PR TITLE
HA: Set Optional ordering for Keystone startup

### PR DESCRIPTION
**Pull request by dirkmueller**
Keystone depends on HAProxy, Database and RabbitMQ to be
started first.

### CODE REVIEW

File: chef/cookbooks/keystone/recipes/ha.rb [hunk 1]
```diff
@@ -65,5 +65,12 @@
     only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
   end
 
+  pacemaker_order "o-#{service_name}" do
+    ordering "g-haproxy g-postgresql g-rabbitmq cl-#{service_name}"
```
Comment (vuntz) [inline chef/cookbooks/keystone/recipes/ha.rb]: That's not correct:
- it doesn't depend on haproxy
- postgresql / rabbitmq might not be on this cluster
- it doesn't depend on rabbitmq (afaik; it's just used for notifications)
- there might not be a `g-postgresql` resource (when drbd is used)

Comment (vuntz) [inline chef/cookbooks/keystone/recipes/ha.rb]: Something like this should be enough:

```ruby
pacemaker_order "o-#{service_name}" do
  score "Mandatory"
  ordering "postgresql cl-#{service_name}"
  action [ :create ]
  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
  only_if "crm configure show postgresql"
end
```

But it's very ugly, and we wanted to write a helper for this.

Comment (vuntz) [inline chef/cookbooks/keystone/recipes/ha.rb]: Btw, with the order, we should start `cl-keystone` in a notification after the creation of the order.

**Comment by vuntz** [comment]
This depends on https://github.com/crowbar/barclamp-pacemaker/pull/179

File: chef/cookbooks/keystone/recipes/ha.rb [hunk 2]
```diff
@@ -65,5 +65,12 @@
     only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
   end
 
+  crowbar_pacemaker_order_if_exists "o-#{service_name}" do
```
Comment (vuntz) [inline chef/cookbooks/keystone/recipes/ha.rb]: Sorry, I renamed to crowbar_pacemaker_order_only_existing (which makes more sense)

File: chef/cookbooks/keystone/recipes/ha.rb [hunk 3]
```diff
@@ -65,5 +65,12 @@
     only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
   end
 
+  crowbar_pacemaker_order_only_existing "o-#{service_name}" do
+    ordering "postgresql rabbitmq cl-#{service_name}"
+    score "Mandatory"
+    action [ :create ]
```
Comment (aspiers) [inline chef/cookbooks/keystone/recipes/ha.rb]: Don't need `[ ]` here.

**Comment by aspiers** [comment]
+1 other than `[ ]` nitpick.

**Comment by aspiers** [comment]
Are we targeting this for GM?  Would require proper QA.

File: chef/cookbooks/keystone/recipes/ha.rb [hunk 4]
```diff
@@ -65,5 +65,12 @@
     only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
   end
 
+  crowbar_pacemaker_order_only_existing "o-#{service_name}" do
+    ordering "postgresql rabbitmq cl-#{service_name}"
+    score "Mandatory"
```
Comment (vuntz) [inline chef/cookbooks/keystone/recipes/ha.rb]: I'm thinking it should be `Optional`, not `Mandatory` as this is really an issue on startup. Keystone is able to recover from a db that fails after startup.

File: chef/cookbooks/keystone/recipes/ha.rb [hunk 5]
```diff
@@ -65,5 +65,12 @@
     only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
   end
 
+  crowbar_pacemaker_order_only_existing "o-#{service_name}" do
```
Comment (vuntz) [inline chef/cookbooks/keystone/recipes/ha.rb]: Should probably be `o-cl-#{service_name}`

**Comment by vuntz** [comment]
Submitted https://github.com/dirkmueller/barclamp-keystone/pull/1 for my fixes.

**Comment by vuntz** [comment]
+1

**Pull request merged**